unicorn/include/cocaine/idl/unicorn.hpp

```diff
@@ -0,0 +1,267 @@
+/*
+* 2015+ Copyright (c) Anton Matveenko <antmat@yandex-team.ru>
+* All rights reserved.
+*
+* This program is free software; you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation; either version 2 of the License, or
+* (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+*/
+
+#ifndef COCAINE_UNICORN_SERVICE_INTERFACE_HPP
+#define COCAINE_UNICORN_SERVICE_INTERFACE_HPP
+
+#include "cocaine/unicorn/path.hpp"
+#include "cocaine/unicorn/value.hpp"
+
+#include <cocaine/rpc/protocol.hpp>
+#include <boost/mpl/list.hpp>
+#include <vector>
+
+namespace cocaine { namespace io {
+
+struct unicorn_tag;
+struct unicorn_final_tag;
+struct unicorn_locked_tag;
+
+using namespace cocaine::unicorn;
+
+struct unicorn {
+    struct create {
+        typedef unicorn_tag tag;
+
+        static const char* alias() {
+            return "create";
+        }
+
+        /**
+        * Create a node if it does not exist.
+        *
+        * path_t - path to change.
+        * value_t - value to write in path
+        **/
+        typedef boost::mpl::list<
+            path_t,
+            value_t
+        > argument_type;
+
+        /**
+        * true if node was created. Error on any kind of error
+        */
+        typedef option_of<
+            bool
+        >::tag upstream_type;
+
+        typedef unicorn_final_tag dispatch_type;
+    };
+
+    struct put {
+        typedef unicorn_tag tag;
+
+        static const char* alias() {
+            return "put";
+        }
+
+        /**
+        *  put command implements cas behaviour. It accepts:
+        *
+        * path_t - path to change.
+        * value_t - value to write in path
+        * version_t - version to compare with. If version in zk do not match - error will be returned.
+        *   -1 indicates that version check is not performed (forced put).
+        **/
+        typedef boost::mpl::list<
+            path_t,
+            value_t,
+            version_t
```

Comment:
I see magic here. Is `-1`value described somewhere inside Zookeeper. You can use optional value to indicate that there is no need to check a version number.